onnxruntime/core/optimizer/selectors_actions/helpers.h

```diff
@@ -0,0 +1,225 @@
+
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(NodesToOptimize);
+
+ private:
+  Node* GetNode(int index, bool required) const {
+    Node* node = nullptr;
+    ORT_ENFORCE(static_cast<size_t>(index) < nodes_.size() &&
+                ((node = nodes_[index]) != nullptr || !required));
+
+    return node;
+  }
+
+  // if last input is variadic, how many additional nodes are there for this input?
+  // first one is included in num_inputs_
+  int num_extra_variadic_inputs_{0};
+  int num_extra_variadic_outputs_{0};
+  std::vector<Node*> nodes_;
+};
+
+// Helper to build a NodesToOptimize instance
+// Use in selector to incrementally add pieces
+struct NodesToOptimizeBuilder {
+  std::vector<Node*> input_nodes;
+  Node* target_node{nullptr};
+  std::vector<Node*> output_nodes;
+  int num_input_defs{-1};
+  int num_output_defs{-1};
+
+  std::unique_ptr<NodesToOptimize> Build() {
+    ORT_ENFORCE(target_node != nullptr, "A target node must be set.");
+    return std::make_unique<NodesToOptimize>(input_nodes, *target_node, output_nodes, num_input_defs, num_output_defs);
+  }
+};
+
+//
+// Action helpers
+//
+
+enum class ArgType { kInput,
+                     kOutput };
+
+// struct to define the location of an input or output definition for a Node
+struct InOutDefSlot {
+  ArgType in_out;
+  int idx;  // idx of -1 means 'all' if a source, or 'end' if appending to a target
+};
+
+// Helper to define moving a value from one node to another
+struct ValueMoveInfo {
+  // simple 1:1 copy
+  ValueMoveInfo(InOutDefSlot src_slot_in, InOutDefSlot dest_slot_in)
+      : src_slot(src_slot_in), dest_slot(dest_slot_in) {}
+
+  // copy all from source to destination
+  ValueMoveInfo(ArgType src_slot_type, ArgType dest_slot_type, bool is_optional = false)
+      : src_slot{src_slot_type, -1},
+        dest_slot{dest_slot_type, -1},
+        copy_all{true},
+        append{true},
+        optional{is_optional} {
+  }
+
+  // append single value (may be variadic) from source to destination
+  ValueMoveInfo(InOutDefSlot src_slot_in, ArgType dest_slot_type, bool is_optional = false)
+      : src_slot(src_slot_in),
+        dest_slot{dest_slot_type, -1},
+        copy_all{false},
+        append{true},
+        optional{is_optional} {}
+
+  InOutDefSlot src_slot;
+  InOutDefSlot dest_slot;
+  bool copy_all{false};  // ignore src_slot.idx and copy all values
+  bool append{false};    // ignore dest_slot.idx and append to existing values
+  bool optional{false};  // optional copy that can be skipped if source node is missing
+
+ private:
+  ValueMoveInfo() = default;
+};
+
+// info to move a value between a source node in the selected_nodes and the target/replacement node.
+struct NodeAndMoveInfo {
+  NodesToOptimize::NodeLocation src_node;
+  ValueMoveInfo value_move_info;
+};
+
+// helpers for moving inputs/outputs and their edges between nodes
+// if there are optional nodes (e.g. bias input to Conv), `required` can be set to false to ignore missing nodes.
```

Comment:
>`required` [](http://example.com/codeflow?start=58&length=10)

is this referring to an old parameter?

Reply:
Yeah. the 'required' aspect became internal to NodesToOptimize.